src/components/verification-grid/verification-grid.ts

```diff
@@ -0,0 +1,1364 @@
+    // start caching client side from the start of the next page
+    const pagesToClientCache = this.pagedItems + this.gridSize;
+
+    const PagesToCacheServerSide = 10;
+    const targetServerCacheHead = pagesToClientCache + this.gridSize * PagesToCacheServerSide;
+
+    this.cacheClient(pagesToClientCache);
+    if (!this.serverCacheExhausted) {
+      this.cacheServer(targetServerCacheHead);
+    }
+  }
+
+  //#endregion
+
+  //#region DownloadingResults
+
+  private canDownloadResults(): boolean {
+    return this.decisions.length > 0;
+  }
+
+  // TODO: clean up this function
+  // TODO: there is a "null" in additional tags (if none)
+  private downloadResults(): void {
+    if (!this.dataSource) {
+      throw new Error("No input data source found");
+    }
+
+    // since we do not know the input format of the provided csv or json files
+    // it is possible for users to input a csv file that already has a column
+    // header of "confirmed" or "additional-tags"
+    // to prevent column name collision, we prepend all the fields that we add
+    // to the original data input with "oe"
+    const columnNamespace = "oe-";
+
+    let formattedResults = "";
+    const fileFormat = this.dataSource.fileType ?? "json";
+
+    const results = this.decisions.map((wrapper: DecisionWrapper) => {
+      // the decisions array is a list of flattened DecisionWrapper's
+      // so that each decision contains only one classification
+      const decision = wrapper.decisions.at(0) as Classification;
+
+      const subject = wrapper.subject;
+      const tag = decision.tag?.text ?? "";
+      const confirmed = decision.confirmed;
+      const additionalTags = wrapper.additionalTags;
+
+      return {
+        ...subject,
+        [`${columnNamespace}tag`]: tag,
+        [`${columnNamespace}confirmed`]: confirmed,
+        [`${columnNamespace}additional-tags`]: additionalTags,
+      };
+    });
+
+    if (fileFormat === "json") {
+      formattedResults = JSON.stringify(results);
+    } else if (fileFormat === "csv") {
+      formattedResults = new Parser().parse(results);
+    } else if (fileFormat === "tsv") {
+      formattedResults = new Parser({ delimiter: "\t" }).parse(results);
+    }
+
+    // create a media type as defined by e.g. application/json, text/csv, etc.
+    // https://www.iana.org/assignments/media-types/media-types.xhtml
+    const topLevelType = fileFormat === "json" ? "application" : "text";
+    const mediaType = `${topLevelType}/${fileFormat}`;
+    const blob = new Blob([formattedResults], { type: mediaType });
+    const url = URL.createObjectURL(blob);
+
+    // TODO: we should just be able to use the file download API
+    // TODO: probably apply a transformation to arrays in CSVs (use semi-columns as item delimiters)
+    const a = document.createElement("a");
+    a.href = url;
+    a.download = `verified-${this.dataSource.fileName}`;
+    a.click();
+    URL.revokeObjectURL(url);
+  }
+
+  //#endregion
```

Comment:
definitely should be in a separate class - probably in the one of the data sources

Reply:
Moved to data-source.ts